# COMP3411/9418 21T0 Assignment 1

- Lecturer: Anna Trofimova
- School of Computer Science and Engineering, UNSW Sydney
- Last Update 13th January at 18:00pm, 2021
$$
% macros
\newcommand{\indep}{\perp \!\!\!\perp}
$$

## Student:
Jie Mei, z5173405

### Submission
This interactive notebook contains the instructions to complete assignment 1; You should submit this notebook with the code and answers in one single file in .ipybn format with the name assignment1.ipybn. **Write your name and zID in the cell above** (to edit the markdown text double-click the cell).

There is a maximum file size cap of 5MB, so make sure your submission does not exceed this size. The submitted notebook should contain all your source code and answers. You can add new cells and use markdown text to organise and explain your implementation/answer.

Submit your files using give. On a CSE Linux machine, type the following on the command-line:

The submission deadline is **22nd January at 11.59pm, 2021.**

### Late Submission Policy 
The penalty is set at 20% per late day. This is ceiling penalty, so if your submission is marked 60/100 and it was submitted two days late, you still get 60/100. If you submit 5 days later, then the penalty is 100% and your mark will be 0.

### Plagiarism
This is an individual assignment. Remember that **all** work submitted for this assignment must be your own **individual** work and no code
sharing or copying is allowed. You may use code from the Internet only with suitable attribution
of the source in your program. **Do not use public code repositories as your code might be copied.** Keep in mind that sharing parts of assignment solutions is a form of plagiarism. All submitted assignments will be run through plagiarism detection software to detect similarities to other submissions. You should carefully read the UNSW policy on academic integrity and plagiarism. 

### **Note: if your life is falling apart and you are tempted to plagiarize as to not fail the assignment, please send me an email instead. I would rather grant you an extention than report you to the Student Conduct and Integrity Unit**.

### Description

This assignment consists of two problems presented in a puzzle game context. For the first problem, you will be implementing a path search problem along with common search algorithms and evaluating their properties. For the second problem, you will be exploring how to define and solve a constraint satisfaction problem. To implement the solutions, you have to use AIPython code that provides you with many implemented classes and examples of their use.  

When working with a Jupyter notebook, you can edit the \*.py files either in the Jupyter interface (in your browser) or with your favorite editor (e.g., PyCharm). Whenever you save a \*.py file, the notebook will reload their content directly.

The libraries that you can use (and need) in this assignment are  the following (run the cell below to import the libraries):

In [1]:
import sys
sys.path.insert(0, "aipython")
import time
import numpy as np
from aipython.searchGeneric import *
from aipython.searchProblem import *
from aipython.cspProblem import *
from aipython.cspSearch import *

*Tip: If you placed the notebook correctly it should not give you any errors.*

*If you think that there are other libraries you might need then send an email to the lecturer-in-charge to confirm.*

### Part 1 - Heuristics [5 marks]

This part of the assignment is based on a popular puzzle, Game of Fifteen, also known as 15-puzzle. The game consists of 15 numbered tiles on a four by four grid where one tile is missing. To solve the puzzle, the tiles must be moved so that they are ordered from 1 to 15. The goal state of the puzzle is defined as a multidimensional array of digits#, where 0 represents the missing tile, as follow:

In [2]:
goal = [[1, 2, 3, 4], 
        [5, 6, 7, 8], 
        [9, 10, 11, 12], 
        [13, 14, 15, 0]]

#### Task 1.1
Edit the code below to implement a search problem class representing the Game of Fifteen that uses heuristic based on Manhattan Distance (h1). The class must be implemented by extending the class Search\_problem. 

Manhattan Distance between two points $p1$ at $(x1, y1)$ and $p2$ at $(x2, y2)$ is $d_{M}(p1, p2) = |x1 - x2| + |y1 - y2|$

*Tip: If you have problems understanding how to overwrite methods take a look at the implementation of Search_problem_from_explicit_graph class in searchProblem.py*


In [3]:
class GameFifteenProblem(Search_problem):
    
    def __init__(self, start, goal):
        start = np.array(start)
        goal = np.array(goal)
        self.start = start
        self.goal = goal
        return
    def start_node(self):
        """Returns the start node"""
        
        return self.start
    
    def is_goal(self, node):
        """Returns True if the node is the goal, otherwise False"""
        np_node = np.array(node)
        np_goal = np.array(self.goal)
        if (np_node != np_goal).any():
            return False
        else:
            return True
    
    def neighbors(self, node):
        """Returns a list of the arcs for the neighbors of node, for example: 
        return [Arc(node, to_neighbor_node1, cost1), Arc(node, to_neighbor_node2, cost2)]"""
        res = []
        np_node = np.array(node)
        zero_point_tuple = np.where(np_node == 0)
        x = int(zero_point_tuple[0])
        y = int(zero_point_tuple[1])
        if x - 1 in range(4):
            top = np_node.copy()
            top[x][y] = top[x - 1][y]
            top[x - 1][y] = 0
            top = top.tolist()
            tmp = Arc(node,top)
            res.append(tmp)
        if x + 1 in range(4):
            down = np_node.copy()
            down[x][y] = down[x + 1][y]
            down[x + 1][y] = 0
            down = down.tolist()
            tmp = Arc(node,down)
            res.append(tmp)
        if y - 1 in range(4):
            left = np_node.copy()
            left[x][y] = left[x][y - 1]
            left[x][y - 1] = 0
            left = left.tolist()
            tmp = Arc(node,left)
            res.append(tmp)
        if y + 1 in range(4):
            right = np_node.copy()
            right[x][y] = right[x][y + 1]
            right[x][y + 1] = 0
            right = right.tolist()
            tmp = Arc(node,right)
            res.append(tmp)
        return res
    
    def heuristic(self, node):
        """Returns the heuristic value of the node 
        based on the Manhattan distance"""
        res = 0
        np_node = np.array(node)
        for i in range(4):
            for j in range(4):
                tmp = self.goal[i][j] 
                if tmp != 0:
                    point = np.where(np_node == tmp)
                    x = int(point[0])
                    y = int(point[1])
                    distance = abs(x - i) + abs(y - j)
                    res += distance
        return res


To validate the correctness of the problem class use the A* searcher algorithm (from searchGeneric.py) to find a solution. 

*Tip: The cost of the solution should be 9.*

In [4]:
start = [[1, 2, 3, 4], 
         [9, 5, 6, 7], 
         [10, 11, 8, 0], 
         [13, 14, 15, 12]]
puzzle = GameFifteenProblem(start, goal)
searcher = AStarSearcher(puzzle)
solution = searcher.search()
print('Cost: ',  solution.cost)

11 paths have been expanded and 23 paths remain in the frontier
Cost:  9


#### Task 1.2 
Implement search problem classes representing the Game of Fifteen that use heuristics based on Euclidean Distance (h2) and the number of the inversions of the permutation (h3). The classes must be implemented by extending the class GameFifteenProblem.

Euclidean distance between two points $p1$ at $(x1, y1)$ and $p2$ at $(x2, y2)$ is $d_{E}(p1, p2) =  \sqrt{(x1-x2)^{2}+ (y1-y2)^{2}}$

An inversion of a permutation $(t_{1},t_{2},...,t_{n})$ of the elements in a finite n-element set of positive integers $A=\{1,2,...,n\}$ is the pair $(t_{j},t_{k})$, where $j<k$ and $t_{j}>t_{k}$.

$\sum_{j=1}^{N} \sum_{k=j}^{N}    \{t_{j} > t_{k}: 1, otherwise: 0\}$, where $N$ is the total number of elements and $t_{i}$ is the value of i-th element.

In [5]:
class GameFifteenProblemEuclidean(GameFifteenProblem):
    def __init__(self, start, goal):
        (super().__init__(start, goal))

    def heuristic(self, node):
        """Returns the heuristic value of the node 
        based on the Euclidean distance"""
        res = 0
        np_node = np.array(node)
        for i in range(4):
            for j in range(4):
                tmp = self.goal[i][j] 
                if tmp != 0:
                    point = np.where(np_node == tmp)
                    x = int(point[0])
                    y = int(point[1])
                    distance = np.sqrt(np.square(x-i) + np.square(y-j))
                    res += distance
        return res

In [6]:
class GameFifteenProblemInversions(GameFifteenProblem):
    def __init__(self, start, goal):
        (super().__init__(start, goal))

    def heuristic(self, node):
        """Returns the heuristic value of the node 
        based on the sum of the inversion number of a permutation"""
        res = 0
        np_node = np.array(node)
        np_flatten = np_node.flatten()
        node_list = np_flatten.tolist()
        for i in range(16):
            j = i
            for j in range(j,16):
                if node_list[j] != 0 and node_list[i] > node_list[j]:
                    res += 1
                
        return res

#### Task 1.3
Run A* Search algorithm with every heuristic for the following three start states:

In [7]:
# optimal path cost: 14
start14 = [[1, 2, 8, 3],
           [5, 6, 7, 4],
           [9, 15, 14, 11],
           [13, 10, 12, 0]]

# optimal path cost: 17
start17 = [[1, 3, 6, 4],
           [5, 2, 8, 14],
           [9, 15, 7, 0],
           [13, 10, 12, 11]]

# optimal path cost: 23
start23 = [[1, 3, 6, 4],
           [5, 8, 15, 14],
           [9, 2, 7, 0],
           [13, 10, 12, 11]]

#start_time = time.time()
puzzle = GameFifteenProblem(start14, goal)
searcher = AStarSearcher(puzzle)
solution = searcher.search()
print('Cost: ',  solution.cost)
#end_time = time.time()
#print("Time: ",end_time - start_time)

45 paths have been expanded and 85 paths remain in the frontier
Cost:  14


In [8]:
puzzle = GameFifteenProblemEuclidean(start23, goal) 
#print(puzzle.heuristic(start23))
searcher = AStarSearcher(puzzle)
solution = searcher.search()
print('Cost: ',  solution.cost)

10083 paths have been expanded and 22192 paths remain in the frontier
Cost:  23


In [9]:
puzzle = GameFifteenProblemInversions(start14, goal) 
#print(puzzle23.heuristic(start23))
searcher = AStarSearcher(puzzle)
solution = searcher.search()
print('Cost: ',  solution.cost)

30 paths have been expanded and 62 paths remain in the frontier
Cost:  14


In each case record in the tables below the heuristic values for the start states, the number of the expanded nodes, and the costs of the solutions.

| Heuristic | h-value: start14 | h-value: start17 | h-value: start23 |
|-----------|------------------|------------------|------------------|
|Manhattan  | 12                | 17                | 19                | 
|Euclidean  | 9.65                | 13.5                | 16.1                | 
|Inversions | 20                | 23                | 29                | 

| Heuristic | expanded: start14 | expanded: start17 | expanded: start23 |
|-----------|-------------------|-------------------|-------------------|
|Manhattan  | 45                 | 28                 | 617                 | 
|Euclidean  | 196                 | 309                 | 10083                 |  
|Inversions | 30                 | 1029                 | 87774                 |  

| Heuristic | path cost: start14 | path cost: start17 | path cost: start23 |
|-----------|--------------------|--------------------|--------------------|
|Manhattan  | 14                  | 17                  | 23                  | 
|Euclidean  | 14                  | 17                  |23                  |
|Inversions | 14                  | 17                  | 27                  |

Comment on the performance of the A* search algorithm with each heuristic based on the results in the tables. Explain which heuristic is better/worse and why.

### Part 2:  Search algorithms [7 marks]

*Tip: If you have problems understanding how to overwrite methods take a look at the implementation of AStarSearcher and  Searcher classes in searchGeneric.py*

*Tip: To initialize the frontier think of the type of structure used by the algorithm to store generated nodes. If it uses a priority queue use FrontierPQ from serachGeneric.py*

#### Task 2.1
Implement a class that performs Breadth First Search by extending class Searcher.

In [10]:
class BreadthFirstSearcher(Searcher):

    def __init__(self, problem):
        super().__init__(problem)

    """ Initializes the forontier """

    def initialize_frontier(self):
        self.frontier = []


    """ Returns True if there are no more nodes to expand """

    def empty_frontier(self):
        return self.frontier == []

    """ Adds the path to the forontier """

    def add_to_frontier(self, path):
        self.frontier.append(path)

    """returns (next) path from the problem's start node
        to a goal node. """

    def search(self):
        while not self.empty_frontier():
            if len(self.frontier) > 1000000:
                return "Out of Memory"
            path = self.frontier.pop(0)
            self.num_expanded += 1
            for neighs in self.problem.neighbors(path.end()):
                if self.problem.is_goal(neighs.to_node):
                    self.display(1, self.num_expanded, "paths have been expanded and",
                                 len(self.frontier), "paths remain in the frontier")
                    path = Path(path, neighs)
                    self.solution = path
                    return path
                self.add_to_frontier(Path(path, neighs))



#### Task 2.2

Implement a class that performs Iterative Deepening Search by extending class Searcher.

In [11]:
class IterativeDeepeningSearcher(Searcher):

    def __init__(self, problem):
        super().__init__(problem)

    """ Initializes the forontier """

    def initialize_frontier(self):
        self.frontier = []

    """ Returns True if there are no more nodes to expand """

    def empty_frontier(self):
        return self.frontier == []

    """ Adds the path to the forontier """

    def add_to_frontier(self, path):
        self.frontier.append(path)

    def search(self):
        depth = -1
        while True:
            depth += 1
            while not self.empty_frontier():
                path = self.frontier.pop()
                self.num_expanded += 1
                if self.problem.is_goal(path.end()):
                    self.display(1, self.num_expanded, "paths have been expanded and",
                                 len(self.frontier), "paths remain in the frontier")
                    self.solution = path
                    return path
                if path.cost > depth:
                    continue
                for neighs in reversed(list(self.problem.neighbors(path.end()))):
                    self.add_to_frontier(Path(path, neighs))
            self.initialize_frontier()
            self.add_to_frontier(Path(self.problem.start_node()))


#### Task 2.3
Implement a class that performs Iterative Deepening A* Search by extending class Searcher.

In [12]:
class IterativeDeepeningAStarSearcher(Searcher):
    """ Initializes the forontier """

    def initialize_frontier(self):
        self.frontier = []

    """ Returns True if there are no more nodes to expand """

    def empty_frontier(self):
        return self.frontier == []

    """ Adds the path to the forontier """

    def add_to_frontier(self, path):
        self.frontier.append(path)

    def search(self):
        thredshold = self.problem.heuristic(self.problem.start_node())
        while True:
            min_value = float("inf")
            while not self.empty_frontier():
                path = self.frontier.pop()
                self.num_expanded += 1
                if self.problem.is_goal(path.end()):
                    self.display(1, self.num_expanded, "paths have been expanded and",
                                 len(self.frontier), "paths remain in the frontier")
                    self.solution = path
                    return path
                fn = path.cost + self.problem.heuristic(path.end())
                if fn > thredshold:
                    if fn < min_value:
                        min_value = fn
                    continue
                for neighs in reversed(list(self.problem.neighbors(path.end()))):
                    self.add_to_frontier(Path(path, neighs))

            thredshold = min_value
            self.initialize_frontier()
            self.add_to_frontier(Path(self.problem.start_node()))


#### Task 2.4
Implement a class that performs Uniform Cost Search by extending class Searcher.

In [13]:
class UniformCostSearcher(Searcher):
    """ Initializes the forontier """
    def initialize_frontier(self):
        self.frontier = FrontierPQ()

    """ Returns True if there are no more nodes to expand """
    def empty_frontier(self):
        return self.frontier.empty()

    """ Adds the path to the forontier """
    def add_to_frontier(self, path):
        value = path.cost
        self.frontier.add(path, value)

#### Task 2.5

Run Breadth First Search (BFS), Iterative Deepenining Search (IDS), Iterative Deepening A* Search (IDA\*S), and Uniform Cost Search (UCS) algorithms on each of the 5 start states:

In [14]:
# optimal path cost: 10
start10 = [[2, 3, 7, 4],
           [1, 6, 11, 8],
           [5, 10, 0, 12],
           [9, 13, 14, 15]]

# optimal path cost: 24
start24 = [[2, 7, 11, 4],
           [6, 3, 12, 0],
           [1, 5, 15, 8],
           [9, 10, 13, 14]]

# optimal path cost: 30
start30 = [[2, 7, 11, 4],
           [6, 3, 12, 0],
           [1, 5, 15, 14],
           [9, 10, 8, 13]]

# optimal path cost: 36
start36 = [[7, 11, 12, 4],
           [2, 3, 14, 1],
           [6, 5, 13, 8],
           [9, 10, 15, 0]]

# optimal path cost: 41
start41 = [[7, 11, 12, 4],
           [2, 3, 8, 14],
           [10, 0, 5, 1],
           [6, 9, 13, 15]]

# your code

In [15]:
start_time1 = time.time()
puzzle = GameFifteenProblem(start24, goal)
searcher = BreadthFirstSearcher(puzzle)
solution = searcher.search()
if time.time() - start_time1 > 300:
    searcher.search().exit()
    print("Out of Time")
elif solution == "Out of Memory":
    print("Out of Memory")
else:
    print('Cost: ',  solution.cost)

Out of Memory


In [16]:
start_time2 = time.time()
puzzle = GameFifteenProblem(start41, goal)
searcher = IterativeDeepeningAStarSearcher(puzzle)
solution = searcher.search()
if time.time() - start_time2 > 300:
    searcher.search().exit()
    print("Out of Time")
elif solution == "Out of Memory":
    print("Out of Memory")
else:
    print('Cost: ',  solution.cost)

2000319 paths have been expanded and 43 paths remain in the frontier
Cost:  41


In each case, record in the table below the number of nodes generated during the search. If the algorithm runs out of memory or the number of nodes in the frontier exceeds 1 million items, just write “Mem” in your table. If the code runs for five minutes without producing output, terminate the process and write “Time” in your table.

*Tip: To edit the table double click the cell below.*

| Algorithm | 10      | 24      | 30      | 36      | 41      |
|-----------|---------|---------|---------|---------|---------|
|BFS        | #       | #       | #       | #       | #       |
|A*         | #       | #       | #       | #       | #       |
|IDS        | #       | #       | #       | #       | #       |
|IDA\*S     | #       | #       | #       | #       | #       |
|UCS        | #       | #       | #       | #       | #       |

Discuss the time and space efficiency of these four algorithms, comment on the results in the table.

### Part 3: Deceptive Starting States [3 marks]

#### Task 3.1 

Run IDA* on the starting states below and report the number of the expanded nodes.

In [26]:
start37_1 = [[7, 11, 12, 4],
             [2, 3, 14, 1],
             [6, 5, 13, 0],
             [9, 10, 15, 8]]

start37_2 = [[7, 11, 12, 4],
             [2, 3, 8, 14],
             [6, 0, 1, 15],
             [9, 5, 10, 13]]
# your code

#### Task 3.2

Explain why there is a difference between the number of the expanded nodes for these starting states if their costs of the optimal paths are the same.

### Part 4:  Constraint Satisfaction Problem [5 marks]

This part of the assignment is based on another puzzle called Sudoku. The game consists of 9x9 grid with a few cells filled in with digits. The objective of the game is to fill the remaining cells so that each column, each row, and each of the nine 3×3 sub-grids contain all of the digits from 1 to 9.

The start state of the puzzle is defined as a multidimensional array of numbers, where 0 represents empty cells, for example:

In [28]:
grid = [[9, 5, 0, 8, 2, 7, 3, 0, 0],
        [0, 8, 0, 1, 4, 0, 0, 5, 0],
        [0, 1, 0, 5, 9, 0, 0, 0, 0],
        [8, 3, 0, 0, 0, 0, 0, 7, 5],
        [1, 6, 9, 7, 5, 2, 4, 3, 0],
        [0, 7, 0, 0, 8, 0, 0, 6, 0],
        [0, 9, 1, 0, 6, 0, 8, 4, 0],
        [7, 0, 8, 0, 3, 1, 0, 0, 6],
        [6, 2, 0, 4, 7, 8, 0, 9, 0]]

*Tip: If you have problems with understanding how to implement constrains take a look at cspExamples.py and cspExamplesQueens.py*

#### Task 4.1

Write the code that defines constraint(s) and a function grid_to_csp that returns a CSP for a Sudoku puzzle:

In [31]:
# define constraint function(s)
def check_rules(point,np_grid):
    x = int(point[0])
    y = int(point[1])
    np_grid_3_3 = [np_grid[x//3*3+i][y//3*3+j] for i in range(3) for j in range(3)]
    res = set(range(1,10)) - set(np_grid_3_3) - set(np_grid[x]) - set(list(zip(*np_grid))[y])

    return res

def ne(a,b):
    return a != b
          
# function that returns a csp
def grid_to_csp(grid):
    np_grid = np.array(grid)
    zero_point = np.where(np_grid == 0)
    x_array = zero_point[0]
    y_array = zero_point[1]
    point_array = []
    for i in range(len(x_array)):
        point = (x_array[i], y_array[i])
        point_array.append(point)
    domains = {point_array[i]:check_rules(point_array[i],np_grid) for i in range(len(point_array))}
    constraints = []
    for i in range(9):
        column_array = np.where(np_grid[i] == 0)
        tuple_array = []
        for j in column_array[0]:
            tmp = (i, j)
            tuple_array.append(tmp)
        for i in range(len(tuple_array)):
            j = i + 1
            while j < len(tuple_array):
                tmp = Constraint([tuple_array[i], tuple_array[j]], ne)
                constraints.append(tmp)
                j += 1

    for i in range(3):
        for j in range(3):
            array = []
            for point in point_array:
                if (i*3 <= point[0] < (i+1)*3) and (j*3 <= point[1] < (j+1)*3):
                    array.append(point)
            for i in range(len(array)):
                j = i + 1
                while j < len(array):
                    tmp = Constraint([array[i], array[j]], ne)
                    constraints.append(tmp)
                    j += 1

    for i in range(9):
        row_array = np.where(np_grid[...,i] == 0)
        tuple_array = []
        for j in row_array[0]:
            tmp = (j, i)
            tuple_array.append(tmp)
        for i in range(len(tuple_array)):
            j = i + 1
            while j < len(tuple_array):
                tmp = Constraint([tuple_array[i], tuple_array[j]], ne)
                constraints.append(tmp)
                j += 1

    return CSP(domains, constraints)

# csp
csp = grid_to_csp(grid)

#### Task 4.2

Write the code that solves Sudoku csp using a search algorithms of your choice, print the solution.

In [32]:
# your code
path = Searcher(Search_from_CSP(csp)).search()
if path is not None:
    print(path.end())
else:
    print("No solution")

2866 paths have been expanded and 10 paths remain in the frontier
{(7, 3): 9, (4, 8): 8, (5, 6): 2, (2, 8): 2, (7, 7): 2, (0, 7): 1, (1, 6): 6, (2, 5): 6, (5, 8): 1, (1, 2): 7, (3, 3): 6, (5, 5): 9, (7, 6): 5, (6, 3): 2, (1, 5): 3, (3, 6): 9, (2, 2): 3, (5, 0): 5, (8, 6): 1, (3, 2): 2, (2, 6): 7, (8, 2): 5, (7, 1): 4, (5, 2): 4, (6, 0): 3, (1, 0): 2, (0, 8): 4, (6, 5): 5, (3, 5): 4, (2, 7): 8, (5, 3): 3, (6, 8): 7, (2, 0): 4, (1, 8): 9, (8, 8): 3, (3, 4): 1, (0, 2): 6}


#### Task 4.3

Descibe what do nodes represent in the search tree. Explain your choice of the search algorithm and compare it with any other search algorithm implemented in this assignemnt. 